UsbIpServer/RegistryUtils.cs

```diff
@@ -5,45 +5,207 @@
+            }
+        }
+
+        public class PersistedDevice
+        {
+            public string Guid { get; }
+            public string BusId { get; }
+            public string Name { get; }
+            public PersistedDevice(string guid, string busid, string name)
+            {
+                Guid = guid;
+                BusId = busid;
+                Name = name;
+            }
+
+        }
+
+        public static RegistryKey? GetRegistryKey(ExportedDevice device)
+        {
+            var deviceKeyNames = Registry.LocalMachine.CreateSubKey(DevicesRegistryPath).GetSubKeyNames();
+            foreach (var keyName in deviceKeyNames)
+            {
+                var deviceKey = Registry.LocalMachine.CreateSubKey(@$"{DevicesRegistryPath}\{keyName}");
+                if (IsDeviceMatch(deviceKey, device))
+                {
+                    return deviceKey;
+                }
+            }
+
+            return null;
+        }
+
+        public static string? GetRegistryKeyName(ExportedDevice device)
+        {
+            var deviceKeyNames = Registry.LocalMachine.CreateSubKey(DevicesRegistryPath).GetSubKeyNames();
+            foreach (var keyName in deviceKeyNames)
+            {
+                var deviceKey = Registry.LocalMachine.CreateSubKey(@$"{DevicesRegistryPath}\{keyName}");
+                if (IsDeviceMatch(deviceKey, device))
+                {
+                    return keyName;
+                }
+            }
+
+            return null;
+        }
+
+
+
+        public static void SetDeviceAsAttached(ExportedDevice device)
+        {
+            var key = GetRegistryKey(device);
+            if (key != null)
             {
-                Registry.LocalMachine.DeleteSubKey($@"{DevicesRegistryPath}\{busId}");
+                key.CreateSubKey("Attached", true, RegistryOptions.Volatile);
             }
         }
 
-        public static string[] GetRegistryDevices()
+        public static void SetDeviceAsDetached(ExportedDevice device)
         {
-            return Registry.LocalMachine.OpenSubKey(DevicesRegistryPath)?.GetSubKeyNames() ?? Array.Empty<string>();
+            var keyName = GetRegistryKeyName(device);
+            if (keyName != null)
+            {
+                Registry.LocalMachine.DeleteSubKeyTree(@$"{DevicesRegistryPath}\{keyName}\Attached");
+            }
         }
 
-        public static void InitializeRegistry()
+        public static void SetDeviceAddress(ExportedDevice device, string address)
         {
-            var devices = Registry.LocalMachine.OpenSubKey(DevicesRegistryPath, true)
-                ?? throw new UnexpectedResultException($"registry key missing; try reinstalling the product");
-            foreach (var subKeyName in devices.GetSubKeyNames())
+            var key = GetRegistryKey(device);
+            if (key != null)
             {
-                devices.DeleteSubKey(subKeyName);
+                key.SetValue("VIP", address);
```

Comment:
W.r.t. my earlier comment. Why not pass an `IPAddress` here. Fits the "strongly typed parameters" principle. Then the conversion to a more readable format can be done here as well (without a helper, even).

Reply:
Good point. Fixed